lib/middleware/karma.js

```diff
@@ -164,10 +162,14 @@ function createKarmaMiddleware (
           const scriptTags = []
           for (const file of files.included) {
             let filePath = file.path
-            const fileType = file.type || path.extname(filePath).substring(1)
-
-            if (helper.isDefined(fileType) && !FILE_TYPES.includes(fileType)) {
-              log.warn(`Invalid file type (${fileType}), defaulting to js.`)
+            const fileType = file.type || file.detectType()
+
+            if (!FILE_TYPES.includes(fileType)) {
+              log.warn(
+                `Invalid file type (${fileType}), defaulting to js.\n` +
```

Comment:
Maybe ($fileType || 'filename has no .extension')

Reply:
Not sure if it really makes the warning more understandable... Before it was simply saying that `$fileType` is an invalid file type, but now it also introduces a relation between file type and file extension without really explaining it.

How about having different messages depending on whether it was an invalid value in the configuration (this should probably be changed to an error, but in a separate PR) vs Karma failed to guess a valid type from the file extension? See updated PR.

My logic with the error message is to say:
- what happened? - Karma failed to guess a file type
- how to fix it? - specify file type in the configuration file
- don't understand what is it about? - read the documentation link that explains it

Reply:
And my assumption here is that normally people want to specify the `type` in the configuration, not change the extension of the file. E.g. testing `.vue` single file components or https://github.com/karma-runner/karma/pull/3491#issuecomment-621737725.